spec/integration/gds_sso_spec.rb

```diff
@@ -91,4 +91,29 @@
       end
     end
   end
+
+  describe "when there is a failure during authentication" do
+    it "is shown if there is an authentication failure with external provider" do
+      OmniAuth.config.test_mode = true
+      OmniAuth.config.mock_auth[:gds] = :invalid_credentials
+
+      logout
+
+      get root_path
+
+      expect(response).to redirect_to "/auth/gds"
+      follow_redirect!
+
+      expect(response).to redirect_to "/auth/gds/callback"
+      follow_redirect!
+
+      expect(response).to redirect_to "/auth/failure?message=invalid_credentials&strategy=gds"
+    end
+
+    it "has a retry link" do
+      get "/auth/failure?message=invalid_credentials&strategy=gds"
+
+      expect(response.body).to include '<a href="/auth/gds">try again</a>'
+    end
+  end
```

Comment:
```suggestion
  describe "when there is a failure during authentication" do
  	let(:test_mode) { true }
  	let(:login_path) { "/auth/gds" }
  	let(:login_callback_path) { "/auth/gds/callback" }
  	let(:invalid_creds_path) { "/auth/failure?message=invalid_credentials&strategy=gds" }
  	
    it "is shown if there is an authentication failure with external provider" do
      OmniAuth.config.mock_auth[:gds] = :invalid_credentials

      logout

      get root_path

      expect(response).to redirect_to login_path
      follow_redirect!

      expect(response).to redirect_to login_callback_path
      follow_redirect!

      expect(response).to redirect_to invalid_creds_path
    end

    it "has a retry link" do
      get invalid_creds_path

      expect(response.body).to include "<a href='#{login_path}'>try again</a>"
    end
  end
```

Reply:
I'm not sure I see the advantage of this, if we're not going to be changing the values of the `let`s.

Reply:
The benefit is that if we ever change those values, all the tests will fail but you only need to update 1 single line rather than doing find/replace and having multiple line changes.

It might also just be me but I find it easier to understand what the link is and what its for. Kind of like using magic numbers in code but its essentially a random string thats looks like some url.

Reply:
I see. I think if we wanted to make these changes it would make sense to do it as a bigger refactor, that affects the whole file, rather than just this short example group.

Reply:
@aliuk2012 is there a way to assign to a `let` variable without having to create a whole `describe` group?

Reply:
@lfdebrux lets chat about this tomorrow when we pair on the trial user specs. I'm thinking why wouldn't you have a describe/context block to call out a specific state that something might be in.